docs/design/plan-modification.md

```diff
@@ -0,0 +1,543 @@
+    markdownDescription: markdownDescription
+  }
+}
+
+type RequiresReplaceIfFunc func(context.Context, state, config attr.Value) (bool, error)
+
+type RequiresReplaceIfModifier struct {
+  f RequiresReplaceIfFunc
+  description string
+  markdownDescription string
+}
+
+func (r RequiresReplaceIfModifier) Modify(ctx context.Context, req ModifyAttributePlanRequest, resp *ModifyAttributePlanResponse) {
+  resp.RequiresReplace = r.f(ctx, req.State, req.Config)
+}
+
+func (r RequiresReplaceIfModifier) Description(ctx context.Context) string {
+  return r.description
+}
+
+func (r RequiresReplaceIfModifier) MarkdownDescription(ctx context.Context) string {
+  return r.markdownDescription
+}
+```
+
+Provider code:
+
+```go
+type fileResourceType struct{}
+
+// GetSchema returns the schema for this resource.
+func (f fileResourceType) GetSchema(_ context.Context) (schema.Schema, []*tfprotov6.Diagnostic) {
+	return schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"favorite_number": {
+				Type:     types.NumberType,
+				Required: true,
+				PlanModifiers: schema.AttributePlanModifiers{
+				  RequiresReplaceIf(func(ctx context.Context, state, config attr.Value) (bool, error) {
+				    stateVal := state.(types.Number)
+				    configVal := config.(types.Number)
+				  
+				    if !stateVal.Unknown && !stateVal.Null && !configVal.Unknown && !configVal.Null {
+				      if configVal.Value.Cmp(stateVal.Value) > 0 {
+				        return true, nil
+				      }
+				    }
+				    return false, nil
+				  }),
+				  CustomModifier,
+				  // ...
+				},
+			},
+		},
+	}, nil
+}
+```
+
+Here, `CustomModifier` is a user-defined `AttributePlanModifier`.
+
+The `AttributePlanModifier`s in the slice of `PlanModifiers` are executed in order. Note that unlike the `customdiff.All` and `customdiff.Sequence` composition helpers in SDKv2, there is no choice to be made here between executing all helpers, and stopping at the first that "returns true", since the function could be setting the `resp.RequiresReplace` bool _or_ modifying the plan. The `Sequence` behaviour could easily be reimplemented, however, as a helper that accepts `AttributePlanModifiers`.
+
+The fields in the `ModifyAttributePlanRequest` and `ModifyAttributePlanResponse` struct are not the same as those in `ModifyResourcePlanRequest` and `ModifyResourcePlanResponse` from option 1. In particular, it is not possible to change the planned value of _other_ attributes inside an attribute's `PlanModifier`. If it were, it would be possible for two or more attributes to have `PlanModifier`s modifying each other's planned values, with no clear indication of the order in which those operations would be performed. 
+
+Framework documentation should make it clear that `schema.Attribute.PlanModifier` is scoped to a single attribute with no access to other attribute values or API requests. If either of these is needed, provider developers should use `tfsdk.Resource.ModifyPlan()`.
```

Comment:
For my own personal edification, could we theoretically support API requests here as well by extending the interface types (such as a `WithProvider` variant, so developers can strongly type the provider in their own function signatures) or extending the request struct with a `Provider tfsdk.Provider` field (this would unfortunately require type assertions)?